keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,313 @@
+   1. marshalling into JSON in-memory
+   2. normalization creating a copy with shared data-structures
+   3. strict immutability check against the old object, coming from proto assuming it is normalized   
+
+### Future outline sketch: field selection
+
+By default the equality is a strict deep-equal equality on a field. This can be generalized
+to an immutability mode. Here are examples:
+
+- **Default**: **Recursive**: means that none of their fields can change, and new fields are not accepted.
+- **Non-recursive**: For an array or a map, one can not remove or add new items, but existing items can be modified. 
+  For maps this means that the the key stay the same, but values might change. For arrays this means that the length
+  stays the same.
+- **Recursive with allowed addition and deletion**:
+  - New/deleted keys in maps are tolerated.
+  - ~~Appended items or removal of the last item in arrays are tolerated.~~
+    For arrays an identifier field is required, e.g. `.name`.
+- ~~**Mutable**: A field set as mutable could cancel the recursive immutability.~~ probably we don't want this.
+
+In CRDs these modes could be set via `x-kubernetes-immutable-mode: non-recursive/recursive/...`. Inside
+of Golang types the `+immutable` marker could be extended to `// +immutable=non-recursive/recursive/...`.
+
+More complex modes like allowed addition and removal with a custom identifier can be expressed with a
+multi key-value syntax:
+
+```
+// No single field in existing containers can be changed, added or deleted.
+// New containers can be added, existing containers can be removed. Containers
+// are identified by its name.
+// +immutable=addition=allowed,removal=allowed,identifier=.name
+Containers []Container
+```
+
+### Future outline: protobuf
+
+Protobuf encoding identifies unset, empty and null values for slices and maps. In native types
+we use a semantic equality to factor those differences out. In CRs decoding and encoding
+is always faithful with the traditional JSON wire format.
+
+When adding protobuf encoding to CRDs in the future, we have to (without major, non-standard
+effort) identify unset, empty and null for CRs as well. This leads to the idea to use a less
+strict equality in the immutabiltiy checks with that case in mind. But protobuf encoding in
+native types also has a normalization effect, namely posted JSON object are normalized through
+the encoding to protobuf when writing to etcd. 
+
+Hence, it looks sensible to split normalization from immutability equality, and keep a strict 
+deep-equal equality even for protobuf, and potentially native types (if we decide to implement 
+immutability for those).
+
+With the container example we get this, with a sketch of a `+normalize` marker which
+normalizes empty and null to undefined:
+
+```
+// No single field in existing containers can be changed, added or deleted.
+// New containers can be added, existing containers can be removed. Containers
```

Comment:
The "existing containers can be removed" case is interesting. We're not supporting that one with this proposal are we?

Reply:
I think I have to update this. It's a bit tricky with kubebuilder's semantics. `+immutable` should recursively apply to the array, i.e. removal and addition are forbidden.

Reply:
fixed. Had copied a wrong example here.